specifications/environment-management.org

```diff
@@ -0,0 +1,95 @@
+* Environment Management: Scope
+
+Specification of how Snapcraft manages local and remote build
+environments and their lifecycle.
+
+This includes locally and remotely provisioned containers, VMs and build
+services such as Launchpad.
+
+* Terminology
+
+** Snapcraft Build Environment Counter (build-count)
+
+Counter that increments every time Snapcraft creates a new build environment.
+Initialized to zero =0=.  Tracked in the environment-manager datastore.
+
+** Project Instance Identification (project-id)
+
+Unique identifier for project, derived from counter =build-count=. The counter
+will be atomically incremented when claiming an identifier.
+
+** Build Environment Instance Naming (project-instance-name)
+
+The build environment's user-facing name.
+
+Format: =snapcraft-<project-id>-<project-name>-<project-id>=
+
+** "Stale" Build Environment
+
+A build environment that is older than 7 days.
+
+* Build Instance Lifecycle
+
+** Creation
+
+When running =snapcraft= inside a project that requires a build environment,
+Snapcraft shall lookup build environments for user-selected provider and see if
+build-environment already exists and is available for re-use.
+
+If no existing build environment is available, =snapcraft= shall create a new
+one, naming it =snapcraft-<counter>-<project_name>=.
+
+** Re-using Build Environments
+
+When a build environment is going to be re-used, Snapcraft shall first verify
+it:
+
+1. was created by a compatible version of Snapcraft.
+
+2. has an operating system version compatible with `base` (or `build-base`).
+
+3. it is not stale (older than 7 days).
+
+If any of these checks fail, Snapcraft shall automatically clean the environment
+and create a new instance to use.
+
+** Removal
+
+The user can remove a build environment instance using two methods:
+
+- Use =snapcraft clean= from within project directory, using existing behavior.
+
+- Use new command =purge= (or =purge-containers=, or
+  =purge-build-environments=?) from anywhere.
+
+**** Removal via =snapcraft clean=
+
+Cleans build instance matching current project identifier.  Command stays the
+same as current behavior, except with a new check to verify the build instance
```

Comment:
```suggestion
Cleans the build instance matching current project identifier.  Command retains its
current behavior, except with a new check to verify the =project-instance-name=
```

I may have got that right, maybe use the terminology as much as possible since it is defined.

Reply:
Committed this, but needs a little further consideration